src/server.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Functions to interact with jenkins server."""
+
+import logging
+import random
+import time
+import typing
+from pathlib import Path
+
+import ops
+import requests
+from pydantic import BaseModel
+
+import state
+
+logger = logging.getLogger(__name__)
+
+JENKINS_WORKDIR = Path("/var/lib/jenkins")
+AGENT_JAR_PATH = Path(JENKINS_WORKDIR / "agent.jar")
+AGENT_READY_PATH = Path(JENKINS_WORKDIR / "agents/.ready")
+ENTRYSCRIPT_PATH = Path(JENKINS_WORKDIR / "entrypoint.sh")
+
+USER = "jenkins"
+GROUP = "jenkins"
+
+
+class Credentials(BaseModel):
+    """The credentials used to register to the Jenkins server.
+
+    Attrs:
+        address: The Jenkins server address to register to.
+        secret: The secret used to register agent.
+    """
+
+    address: str
+    secret: str
+
+    @classmethod
+    def from_jenkins_slave_interface_dict(
+        cls, server_unit_databag: ops.RelationDataContent
+    ) -> typing.Optional["Credentials"]:
+        """Import server metadata from databag in slave relation.
+
+        Args:
+            server_unit_databag: The relation databag content from slave relation.
+
+        Returns:
+            Metadata if complete values(url, secret) are set. None otherwise.
+        """
+        address = server_unit_databag.get("url")
+        secret = server_unit_databag.get("secret")
+        if not address or not secret:
+            return None
+        return Credentials(address=address, secret=secret)
+
+    @classmethod
+    def from_jenkins_agent_v0_interface_dict(
+        cls, server_unit_databag: ops.RelationDataContent, unit_name: str
+    ) -> typing.Optional["Credentials"]:
+        """Import server metadata from databag in agent relation.
+
+        Args:
+            server_unit_databag: The relation databag content from agent relation.
+            unit_name: The agent unit name.
+
+        Returns:
+            Metadata if complete values(url, secret) are set. None otherwise.
+        """
+        address = server_unit_databag.get("url")
+        secret = server_unit_databag.get(f"{unit_name}_secret")
+        if not address or not secret:
+            return None
+        return Credentials(address=address, secret=secret)
+
+
+def get_credentials(
```

Comment:
Perhaps this function won't be needed if each of the relations has their own event handler?

Reply:
This is moved to be handled in `state.py`, since it seems more suited as a state :D, thank you for the comment!